### PR TITLE
Changes allowing more flexibility with repos and CodeBuild

### PR DESCRIPTION
**Pull request by YakDriver**
Many thing is 3 new variables: CODEBUILD_USE_VERSION, CODEBUILD_GIT_REPO_VAR, and CODEBUILD_GIT_REF_VAR, all optional

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,8 @@ jspm_packages
 
 # Serverless directories
 .serverless
+envvars.sh
```
Comment (lorengordon) [inline .gitignore]: How about putting these files into ./private and having just one entry here for `.private`? That way it works for anyone, regardless of the filenames they choose to use.

File: README.md [hunk 2]
```diff
@@ -46,6 +46,9 @@ export GITHUB_STATUS_CONTEXT="codebuild/pr"  # Tag associated with the status ch
 export GITHUB_BUILD_EVENTS="pr_state"  # Comma-separated string of build events, supports "pr_state" and "pr_comment"
 export GITHUB_BUILD_USERS=""  # Comma-separated string of GitHub users authorized to build from PR comments
 export GITHUB_BUILD_COMMENT="go codebuild go"  # Case-insensitive string that will start a build from a PR comment
+export CB_EXTERNAL_BUILDSPEC=false # If "false", webhook will not have CodeBuild submit source version in buildspec.yml request (useful if in 2 different repos or you want to use the most recent commit)
```
Comment (lorengordon) [inline README.md]: The default should be "true" to retain prior behavior.

Also, I think the description here is too much detail. Should be sufficient to say something like "Set to false if the repo for the CodeBuild project itself is different from the repo where the webhook will be created"

File: README.md [hunk 3]
```diff
@@ -46,6 +46,9 @@ export GITHUB_STATUS_CONTEXT="codebuild/pr"  # Tag associated with the status ch
...
+export CB_GIT_REPO_ENV=REPO_TO_BUILD # Name of env var to pass to CodeBuild with repo triggering build
```
Comment (lorengordon) [inline README.md]: "Name of ENV variable that your CodeBuild project uses to know which repo to clone. For use with CB_EXTERNAL_BUILDSPEC."

File: README.md [hunk 4]
```diff
@@ -46,6 +46,9 @@ export GITHUB_STATUS_CONTEXT="codebuild/pr"  # Tag associated with the status ch
...
+export CB_GIT_REF_ENV=REF_TO_BUILD # Name of env var to pass to CodeBuild with PR# triggering build
```
Comment (lorengordon) [inline README.md]: "Name of ENV variable that your CodeBuild project uses to know which git ref to checkout. For use with CB_EXTERNAL_BUILDSPEC."

File: envvars [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (lorengordon) [inline envvars]: This file should be hidden in .private

File: README.md (hunk 2 above)
Comment (lorengordon) [reply README.md]: Oops, we flipped the logic, "false" is the correct default. "Set to _true_ if ..."

File: handler.js [hunk 6]
```diff
@@ -43,6 +43,14 @@ buildUsers = buildUsers != "" ? buildUsers.split(",") : [];
 var buildComment = getObjectDefault(process.env, "GITHUB_BUILD_COMMENT", "");
 buildComment = buildComment != "" ? buildComment : "go codebuild go";
 
+// if set to true, CodeBuild will not include PR number in request from buildspec from git
```
Comment (lorengordon) [inline handler.js]: "Controls whether the `sourceVersion` parameter is set to the pull request and passed to the CodeBuild job. Set to true if the repo for the CodeBuild project itself is different from the repo where the webhook will be created"

File: handler.js [hunk 7]
```diff
@@ -67,16 +78,37 @@ module.exports.start_build = (event, context, callback) => {
...
       var head = pullRequest.head;
       var base = pullRequest.base;
       var repo = base.repo;
 
       var params = {
-        projectName: process.env.BUILD_PROJECT,
-        sourceVersion: 'pr/' + pullRequest.number
+        projectName: process.env.CB_BUILD_PROJECT
       };
+      if (buildOptions.cbGitRepoEnv && buildOptions.cbGitRefEnv ) {
```
Comment (lorengordon) [inline handler.js]: The logic around this needs a bit of consideration. A git ref is not _required_ when cloning a repo. The clone will simply use the repo's default branch.

Also, what is the rationale for not nesting this test in the `!buildOptions.cbExternalBuildspec` test? I can't think of any reason to use these params separately from that param.

File: serverless.yml [hunk 8]
```diff
@@ -9,14 +9,17 @@ plugins:
 custom:
   stage: dev
   github-webhook:
-    githubStatusContext: codebuild/${env:BUILD_PROJECT}/pr
+    githubStatusContext: codebuild/${env:CB_BUILD_PROJECT}/pr
     githubBuildEvents: pr_state
     githubBuildUsers: ""
     githubBuildComment: go codebuild go
+    cbExternalBuildspec: ""
```
Comment (lorengordon) [inline serverless.yml]: default to "false"

File: serverless.yml [hunk 9]
```diff
@@ -38,6 +41,18 @@ custom:
       Type: String
       Default: ${env:GITHUB_BUILD_COMMENT, self:custom.github-webhook.githubBuildComment}
       Description: For pull request comments, the comment that will initiate a build
+    CbExternalBuildspec:
+      Type: String
+      Default: ${env:CB_EXTERNAL_BUILDSPEC, self:custom.github-webhook.cbExternalBuildspec}
+      Description: If true, CodeBuild will not send "sourceVersion" with request for buildspec (e.g., allowing buildspec to be in an external repo not on same PR number)
```
Comment (lorengordon) [inline serverless.yml]: "Controls whether the sourceVersion parameter is set to the pull request and passed to the CodeBuild job. Set to true if the repo for the CodeBuild project itself is different from the repo where the webhook will be created"

File: serverless.yml [hunk 10]
```diff
@@ -38,6 +41,18 @@ custom:
...
+    CbGitRepoEnv:
+      Type: String
+      Default: ${env:CB_GIT_REPO_ENV, self:custom.github-webhook.cbGitRepoEnv}
+      Description: Name of the env used to pass repo clone URL to CodeBuild (optional, if not set, not passed)
```
Comment (lorengordon) [inline serverless.yml]: "(Optional) Name of ENV variable that your CodeBuild project uses to know which repo to clone. For use with CB_EXTERNAL_BUILDSPEC."

File: serverless.yml [hunk 11]
```diff
@@ -38,6 +41,18 @@ custom:
...
+    CbGitRefEnv:
+      Type: String
+      Default: ${env:CB_GIT_REF_ENV, self:custom.github-webhook.cbGitRefEnv}
+      Description: Name of the env used to pass PR number to CodeBuild (optional, if not set, not passed)
```
Comment (lorengordon) [inline serverless.yml]: "(Optional) Name of ENV variable that your CodeBuild project uses to know which git ref to checkout. For use with CB_EXTERNAL_BUILDSPEC."

**Review by lorengordon — CHANGES_REQUESTED** [review]
Ok, left some comments mostly about refining the descriptions.

I do think the logic testing the git repo and ref envs needs a little more consideration, since only the repo is technically required. Also, I don't see any particular reason to use these variables separately from cb_external_buildspec, so I think the tests can be nested or combined.

**Comment by lorengordon** [comment]
Oh, and please squash/rebase the extraneous commits, most specifically so that any files that are not part of the repo do not appear in the commit history.

**Comment by YakDriver** [comment]
Changes suggested have been incorporated, specifically descriptions of new variables and allowing for ref to be optional if repo is used (already did it that way on the other side, not sure why it was this way on this side).

File: handler.js [hunk 12]
```diff
@@ -67,17 +78,43 @@ module.exports.start_build = (event, context, callback) => {
...
+            params.environmentVariablesOverride = [
+              {
+                name: buildOptions.cbGitRefEnv,
+                type: "PLAINTEXT",
+                value: pullRequest.number + ""
+              },
+            ]
+          }
+        }
+        console.log("Not using the source version! Will likely use most-recently commited buildspec.yml instead. (Use CB_EXTERNAL_BUILDSPEC=true parameter to change.)");
```
Comment (lorengordon) [inline handler.js]: The message here I think can be more straightforward. "Using an external buildspec. Will not pass `sourceVersion` when starting the CodeBuild job. Set CB_EXTERNAL_BUILDSPEC=false to change."

**Review by lorengordon — COMMENTED** [review]
Awesome, just one last change....

**Comment by YakDriver** [comment]
Revised

**Review by lorengordon — APPROVED** [review]

**Pull request merged**